File: backJoon/src/Q1504_특정한_최단_경로/Main.java
package Q1504_특정한_최단_경로;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static class Node {
        // 진행하는 경로
        int next;
        // 비용
        int cost;

        public Node(int next, int cost) {
            this.next = next;
            this.cost = cost;
        }
    }

    static List<Node>[] map;
    static final int INF = (int) 1e9;
    static int N;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());

        map = new ArrayList[N + 1];

        for(int i=1; i<=N; i++) map[i] = new ArrayList<>();

        while(--E>=0) {
            st = new StringTokenizer(br.readLine());
            int prev = Integer.parseInt(st.nextToken());
            int next = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());

            map[prev].add(new Node(next, cost));
            map[next].add(new Node(prev, cost));
        }

        st = new StringTokenizer(br.readLine());
        int nodeA = Integer.parseInt(st.nextToken());
        int nodeB = Integer.parseInt(st.nextToken());

        // 반드시 두 정점 a,b 를 지나는 최단 경로
        // (시작점 -> a) + (a -> b) + (b -> 끝점) 혹은 (시작점 -> b) + (b -> a) + (a -> 끝점)
        int case1 = dijkstra(1, nodeA) + dijkstra(nodeA, nodeB) + dijkstra(nodeB, N);
        int case2 = dijkstra(1, nodeB) + dijkstra(nodeB, nodeA) + dijkstra(nodeA, N);

        if (case1 < 0 || case2 < 0 || case1 >= INF || case2>= INF) System.out.println(-1);
        else System.out.println(Math.min(case1, case2));
    }

    private static int dijkstra(int st, int ed) {
        PriorityQueue<Node> pq = new PriorityQueue<>((n1, n2) -> n1.cost - n2.cost);
        boolean[] visited = new boolean[N+1];
        int[] dist = new int[N+1];
        Arrays.fill(dist, INF);

        dist[st] = 0;
        pq.add(new Node(st, 0));

        while (!pq.isEmpty()) {
            Node curr = pq.poll();
            if (visited[curr.next]) continue;
            visited[curr.next] = true;

            for (Node nNode : map[curr.next]) {
                if (dist[nNode.next] > nNode.cost + curr.cost) {
                    dist[nNode.next] = nNode.cost + curr.cost;
                    pq.add(new Node(nNode.next, dist[nNode.next]));
                }
            }
        }

        return dist[ed];
    }
}

File: backJoon/src/Q1504_특정한_최단_경로/main.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;
int N, E;
struct node {
    int next;
    int dist;
};

struct compare {
    bool operator()(node &n1, node &n2) {
        return n1.dist > n2.dist;
    }
};

vector<node> v[804];
int n1, n2, ans1, ans2;
const int INF = 200000001;

void input() {
    ios::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);
    cin >> N >> E;
    int st, ed, d;
    for (int i = 0; i < E; i++) {
        cin >> st >> ed >> d;
        v[st].push_back({ed, d});
        v[ed].push_back({st, d});
    }
    cin >> n1 >> n2;
}

int dijkstra(int st, int ed) {
    priority_queue<node, vector<node>, compare> pq;
    pq.push({st, 0});
    int dist[N+1];
    fill(dist, dist + N + 1, INF);
    dist[st] = 0;
    while (!pq.empty()) {
        node curr = pq.top();
        pq.pop();
        for (node n: v[curr.next]) {
            if (dist[n.next] > n.dist + curr.dist) {
                dist[n.next] = n.dist + curr.dist;
                pq.push({n.next, dist[n.next]});
            }
        }
    }
    return dist[ed];
}

void solve() {
    ans1 = dijkstra(1, n1) + dijkstra(n1, n2) + dijkstra(n2, N);
    ans2 = dijkstra(1, n2) + dijkstra(n2, n1) + dijkstra(n1, N);
}

void output() {
    cout << (min(ans1, ans2) >= INF || min(ans1, ans2) <= 0 ? -1 : min(ans1, ans2));
}

int main() {
    input();
    solve();
    output();
}